### PR TITLE
Patch to upstream

### PR DESCRIPTION
**Pull request by druzhkov-paul**
For better review experience of #106. This PR not intended to be merged ever.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -112,6 +106,7 @@ jobs:
           export CUDA_HOME=/usr/local/cuda-${CUDA_SHORT}
           export LD_LIBRARY_PATH=${CUDA_HOME}/lib64:${CUDA_HOME}/include:${LD_LIBRARY_PATH}
           export PATH=${CUDA_HOME}/bin:${PATH}
+          export FORCE_CUDA="1"
```
Comment (druzhkov-paul) [inline .github/workflows/build.yml]: Why?

File: .isort.cfg [hunk 2]
```diff
@@ -0,0 +1,8 @@
+[isort]
```
Comment (druzhkov-paul) [inline .isort.cfg]: What's the reason to add this?

File: configs/albu_example/mask_rcnn_r50_fpn_albu_1x_coco.py [hunk 3]
```diff
@@ -31,7 +31,7 @@
                 p=1.0)
         ],
         p=0.1),
-    dict(type='JpegCompression', quality_lower=85, quality_upper=95, p=0.2),
+    dict(type='ImageCompression', quality_lower=85, quality_upper=95, p=0.2),
```
Comment (druzhkov-paul) [inline configs/albu_example/mask_rcnn_r50_fpn_albu_1x_coco.py:34]: Why is it changed?

File: configs/fcos/fcos_center_r50_caffe_fpn_gn-head_4x4_1x_coco.py [hunk 4]
```diff
@@ -0,0 +1,2 @@
+_base_ = './fcos_r50_caffe_fpn_gn-head_4x4_1x_coco.py'
```
Comment (druzhkov-paul) [inline configs/fcos/fcos_center_r50_caffe_fpn_gn-head_4x4_1x_coco.py]: Please remove obsolete (if they are) configs.

File: docs/conf.py [hunk 5]
```diff
@@ -18,17 +18,9 @@
 
 # -- Project information -----------------------------------------------------
 
-project = 'MMDetection'
-copyright = '2018-2020, OpenMMLab'
-author = 'MMDetection Authors'
-version_file = '../mmdet/version.py'
-
-
-def get_version():
```
Comment (druzhkov-paul) [inline docs/conf.py]: I guess this declaration is required.

File: mmdet/apis/test.py [hunk 6]
```diff
@@ -99,7 +99,7 @@ def multi_gpu_test(model, data_loader, tmpdir=None, gpu_collect=False):
             if isinstance(result[0], tuple):
                 result = [(bbox_results, encode_mask_results(mask_results))
                           for bbox_results, mask_results in result]
-        results.extend(result)
+            results.extend(result)
```
Comment (druzhkov-paul) [inline mmdet/apis/test.py]: Is it possible not to introduce this change?

File: mmdet/core/anchor/anchor_generator.py [hunk 7]
```diff
@@ -252,19 +253,20 @@ def single_level_grid_anchors(self,
         """
         feat_h, feat_w = featmap_size
         # convert Tensor to int, so that we can covert to ONNX correctlly
-        feat_h = int(feat_h)
-        feat_w = int(feat_w)
+        #feat_h = int(feat_h)
+        #feat_w = int(feat_w)
```
Comment (druzhkov-paul) [inline mmdet/core/anchor/anchor_generator.py]: Clean up the code, please.

File: mmdet/core/anchor/anchor_generator.py [hunk 8]
```diff
@@ -322,10 +324,10 @@ def single_level_valid_flags(self,
         valid_y = torch.zeros(feat_h, dtype=torch.bool, device=device)
         valid_x[:valid_w] = 1
         valid_y[:valid_h] = 1
-        valid_xx, valid_yy = self._meshgrid(valid_x, valid_y)
+        valid_xx, valid_yy = meshgrid(valid_x, valid_y)
```
Comment (druzhkov-paul) [inline mmdet/core/anchor/anchor_generator.py:322]: Looks like our bug: anchors and their validity flags might be in an inconsistent order. Need a bit more investigation. You may ignore it now.

File: mmdet/core/anchor/anchor_generator.py [hunk 9]
```diff
@@ -252,19 +253,20 @@ def single_level_grid_anchors(self,
...
         shift_x = torch.arange(0, feat_w, device=device) * stride[0]
         shift_y = torch.arange(0, feat_h, device=device) * stride[1]
-
-        shift_xx, shift_yy = self._meshgrid(shift_x, shift_y)
-        shifts = torch.stack([shift_xx, shift_yy, shift_xx, shift_yy], dim=-1)
+        shift_yy, shift_xx = meshgrid(shift_y, shift_x)
+        shifts = torch.stack([shift_xx, shift_yy, shift_xx, shift_yy], dim=-1).view(-1, 1, 4)
+        # open-mmlab/mmdetection suggests the following shifts
+        #shift_xx, shift_yy = self._meshgrid(shift_x, shift_y)
+        #shifts = torch.stack([shift_xx, shift_yy, shift_xx, shift_yy], dim=-1)
```
Comment (druzhkov-paul) [inline mmdet/core/anchor/anchor_generator.py]: Remove these comments please.

File: mmdet/core/bbox/assigners/atss_assigner.py [hunk 10]
```diff
@@ -119,9 +120,11 @@ def assign(self,
...
             end_idx = start_idx + bboxes_per_level
             distances_per_level = distances[start_idx:end_idx, :]
-            selectable_k = min(self.topk, bboxes_per_level)
-            _, topk_idxs_per_level = distances_per_level.topk(
-                selectable_k, dim=0, largest=False)
+            _, topk_idxs_per_level = topk(distances_per_level, self.topk, dim=0, largest=False)
+            # open-mmlab/mmdetection suggests the following topk using
+            #selectable_k = min(self.topk, bboxes_per_level)
+            #_, topk_idxs_per_level = distances_per_level.topk(
+            #    selectable_k, dim=0, largest=False)
```
Comment (druzhkov-paul) [inline mmdet/core/bbox/assigners/atss_assigner.py]: Ditto

File: mmdet/core/post_processing/bbox_nms.py [hunk 11]
```diff
@@ -12,74 +22,86 @@ def multiclass_nms(multi_bboxes,
...
-        nms_thr (float): NMS IoU threshold
-        max_num (int, optional): if there are more than max_num bboxes after
-            NMS, only top max_num will be kept. Default to -1.
-        score_factors (Tensor, optional): The factors multiplied to scores
-            before applying NMS. Default to None.
+        nms_cfg (dict): NMS operation config
+        max_num (int): if there are more than max_num bboxes after NMS,
+            only top max_num will be kept.
+        score_factors (Tensor): The factors multiplied to scores before
+            applying NMS
```
Comment (druzhkov-paul) [inline mmdet/core/post_processing/bbox_nms.py]: Please align these doc strings.

File: mmdet/core/post_processing/bbox_nms.py [hunk 12]
```diff
@@ -1,6 +1,16 @@
+import sys
 import torch
-from mmcv.ops.nms import batched_nms
-
+import torch.onnx.symbolic_helper as sym_help
+from torch.autograd import Function
+from torch.onnx import is_in_onnx_export
+from torch.onnx.symbolic_opset9 import reshape
+from torch.onnx.symbolic_opset10 import _slice
```
Comment (druzhkov-paul) [inline mmdet/core/post_processing/bbox_nms.py]: Please remove unused imports if any.

File: mmdet/core/utils/lr_updater.py [hunk 13]
```diff
@@ -0,0 +1,43 @@
+from mmcv.runner import HOOKS
+from mmcv.runner.hooks.lr_updater import CosineAnnealingLrUpdaterHook, annealing_cos
+
+
+@HOOKS.register_module()
+class CosineAnealingUntilEpochLrUpdaterHook(CosineAnnealingLrUpdaterHook):
+    # TODO: Fix Anealing -> Annealing when the upgrade of MMCV will be possible
```
Comment (druzhkov-paul) [inline mmdet/core/utils/lr_updater.py]: Could you please do that?

File: mmdet/datasets/pipelines/loading.py [hunk 14]
```diff
@@ -380,7 +391,9 @@ def __repr__(self):
         repr_str += f'with_mask={self.with_mask}, '
         repr_str += f'with_seg={self.with_seg}, '
         repr_str += f'poly2mask={self.poly2mask}, '
+        repr_str += f'with_text={self.with_text}, '
         repr_str += f'poly2mask={self.file_client_args})'
+        #repr_str += f'file_client_args={self.file_client_args})'
```
Comment (druzhkov-paul) [inline mmdet/datasets/pipelines/loading.py]: Is it here on purpose?

File: mmdet/datasets/samplers/distributed_sampler.py [hunk 15]
```diff
@@ -20,6 +20,8 @@ def __iter__(self):
             indices = torch.arange(len(self.dataset)).tolist()
 
         # add extra samples to make it evenly divisible
+        while len(indices) < self.total_size:
+            indices += indices[:(self.total_size - len(indices))]
```
Comment (druzhkov-paul) [inline mmdet/datasets/samplers/distributed_sampler.py]: This chunk of code is in conflict (does the same thing) with the one right below. Please leave only one of them. I'd propose to leave the one from upstream.

File: mmdet/datasets/utils.py [hunk 16]
```diff
@@ -127,7 +127,7 @@ def _check_head(self, runner):
         else:
             for name, module in model.named_modules():
                 if hasattr(module, 'num_classes') and not isinstance(
-                        module, (RPNHead, VGG, FusedSemanticHead, GARPNHead)):
+                        module, (RPNHead, VGG, FusedSemanticHead, GARPNHead)) and name != 'backbone':
```
Comment (druzhkov-paul) [inline mmdet/datasets/utils.py:130]: What's the purpose of the change?

File: mmdet/models/dense_heads/fcos_head.py [hunk 17]
```diff
@@ -412,6 +410,34 @@ def _get_points_single(self,
...
+    def _get_points_single(self, featmap_size, stride, dtype, device):
+        h, w = featmap_size
+        x_range = torch.arange(
+            0, w * stride, stride, dtype=dtype, device=device)
+        y_range = torch.arange(
+            0, h * stride, stride, dtype=dtype, device=device)
+        y, x = meshgrid(y_range, x_range)
+        points = torch.stack(
+            (x.reshape(-1), y.reshape(-1)), dim=-1) + stride // 2
+        return points
```
Comment (druzhkov-paul) [inline mmdet/models/dense_heads/fcos_head.py]: What's happening here?

File: mmdet/models/dense_heads/fovea_head.py [hunk 18]
```diff
@@ -120,6 +123,18 @@ def forward_single(self, x):
...
+        for featmap_size in featmap_sizes:
+            x_range = arange(0, featmap_size[1], dtype=dtype, device=device) + 0.5
+            y_range = arange(0, featmap_size[0], dtype=dtype, device=device) + 0.5
+            y, x = meshgrid(y_range, x_range)
+            if flatten:
+                y = y.reshape(-1)
+                x = x.reshape(-1)
+            points.append((y, x))
+        return points
+
```
Comment (druzhkov-paul) [inline mmdet/models/dense_heads/fovea_head.py]: Where does this function used?

File: mmdet/models/detectors/base.py [hunk 19]
```diff
@@ -1,13 +1,18 @@
...
 import numpy as np
+import pycocotools.mask as maskUtils
 import torch
 import torch.distributed as dist
 import torch.nn as nn
+import warnings
+from abc import ABCMeta, abstractmethod
+from collections import OrderedDict
+from contextlib import contextmanager
+from functools import partial
```
Comment (druzhkov-paul) [inline mmdet/models/detectors/base.py:12]: Imports order looks weird.

File: mmdet/models/detectors/base.py [hunk 20]
```diff
@@ -328,10 +401,24 @@ def show_result(self,
...
         if segm_result is not None and len(labels) > 0:  # non empty
             segms = mmcv.concat_list(segm_result)
-            if isinstance(segms[0], torch.Tensor):
-                segms = torch.stack(segms, dim=0).detach().cpu().numpy()
-            else:
-                segms = np.stack(segms, axis=0)
+            #if isinstance(segms[0], torch.Tensor):
+            #    segms = torch.stack(segms, dim=0).detach().cpu().numpy()
+            #else:
+            #    segms = np.stack(segms, axis=0)
```
Comment (druzhkov-paul) [inline mmdet/models/detectors/base.py]: Clean up the code please.

File: mmdet/models/roi_heads/htc_roi_head.py [hunk 21]
```diff
@@ -173,8 +176,11 @@ def _bbox_forward(self, stage, x, rois, semantic_feat=None):
             bbox_semantic_feat = self.semantic_roi_extractor([semantic_feat],
                                                              rois)
             if bbox_semantic_feat.shape[-2:] != bbox_feats.shape[-2:]:
-                bbox_semantic_feat = F.adaptive_avg_pool2d(
-                    bbox_semantic_feat, bbox_feats.shape[-2:])
+                if torch.onnx.is_in_onnx_export():
+                    bbox_semantic_feat = self.avg_pooling(bbox_semantic_feat)
+                else:
+                    bbox_semantic_feat = F.adaptive_avg_pool2d(
```
Comment (druzhkov-paul) [inline mmdet/models/roi_heads/htc_roi_head.py]: Guess these are not equivalent.

File: mmdet/models/roi_heads/mask_heads/fcn_mask_head.py [hunk 22]
```diff
@@ -9,6 +9,8 @@
 
 from mmdet.core import mask_target
 from mmdet.models.builder import HEADS, build_loss
+from mmcv.ops.carafe import CARAFEPack
```
Comment (druzhkov-paul) [inline mmdet/models/roi_heads/mask_heads/fcn_mask_head.py]: Is it used?

File: mmdet/utils/collect_env.py [hunk 23]
```diff
@@ -1,13 +1,84 @@
...
+    env_info['GCC'] = gcc
+
+    env_info['PyTorch'] = torch.__version__
+    env_info['PyTorch compiling details'] = torch.__config__.show()
+
+    env_info['TorchVision'] = torchvision.__version__
+
+    env_info['OpenCV'] = cv2.__version__
+
+    env_info['MMCV'] = mmcv.__version__
```
Comment (druzhkov-paul) [inline mmdet/utils/collect_env.py]:
```suggestion
    """Collect the information of the running environments."""
    env_info = collect_base_env()
```

File: mmdet/utils/collect_env.py [hunk 24]
```diff
@@ -1,13 +1,84 @@
...
+import os.path as osp
+import subprocess
+import sys
+from collections import defaultdict
+from contextlib import suppress
+
+import cv2
+import mmcv
+import torch
+import torchvision
```
Comment (druzhkov-paul) [inline mmdet/utils/collect_env.py]:
```suggestion
from contextlib import suppress

from mmcv.utils import collect_env as collect_base_env
from mmcv.utils import get_git_hash
```

File: requirements/optional.txt [hunk 25]
```diff
@@ -1,6 +1,10 @@
...
 imagecorruptions
+scikit-learn
+ninja
+tensorboard
+imagesize==1.2.0
+PyYAML==5.3.1
+sty==1.0.0b12
 mmlvis
-scipy
-sklearn
```
Comment (druzhkov-paul) [inline requirements/optional.txt]: Why these two are removed?

File: setup.py [hunk 26]
```diff
@@ -156,6 +155,25 @@ def gen_packages_items():
...
+                module='mmdet.ops.roi_align',
+                sources=[
+                    'src/roi_align_ext.cpp',
+                    'src/cpu/roi_align_v2.cpp',
+                ],
+                sources_cuda=[
+                    'src/cuda/roi_align_kernel.cu',
+                    'src/cuda/roi_align_kernel_v2.cu'
+                ]),
+        ],
```
Comment (druzhkov-paul) [inline setup.py]: Let's get rid of these duplicates and move to mmcv.

File: .github/CONTRIBUTING.md [hunk 27]
```diff
@@ -1 +1,52 @@
-We appreciate all contributions to improve MMDetection. Please refer to [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) in MMCV for more details about the contributing guideline.
```
Comment (Ilya-Krylov) [inline .github/CONTRIBUTING.md:1]: I think we can use something quite similar to upstream version.

-----
We appreciate all contributions to improve OTEDetection. Please refer to [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) in MMCV for more details about the contributing guideline.

File: README.md [hunk 28]
```diff
@@ -1,164 +1,42 @@
...
-- [MMClassification](https://github.com/open-mmlab/mmclassification): OpenMMLab image classification toolbox and benchmark.
-- [MMDetection](https://github.com/open-mmlab/mmdetection): OpenMMLab detection toolbox and benchmark.
-- [MMDetection3D](https://github.com/open-mmlab/mmdetection3d): OpenMMLab's next-generation platform for general 3D object detection.
-- [MMSegmentation](https://github.com/open-mmlab/mmsegmentation): OpenMMLab semantic segmentation toolbox and benchmark.
-- [MMAction2](https://github.com/open-mmlab/mmaction2): OpenMMLab's next-generation action understanding toolbox and benchmark.
-- [MMTracking](https://github.com/open-mmlab/mmtracking): OpenMMLab video perception toolbox and benchmark.
-- [MMPose](https://github.com/open-mmlab/mmpose): OpenMMLab pose estimation toolbox and benchmark.
-- [MMEditing](https://github.com/open-mmlab/mmediting): OpenMMLab image and video editing toolbox.
+This repo is currently maintained by Ilya Krylov ([@Ilya-Krylov](https://github.com/Ilya-Krylov)), Pavel Druzhkov ([@druzhkov-paul](https://github.com/druzhkov-paul)) and Alexander Dokuchaev ([@AlexanderDokuchaev](https://github.com/AlexanderDokuchaev)).
```
Comment (Ilya-Krylov) [inline README.md]: @DmitriySidnev please add yourself

File: docker/Dockerfile.cpu [hunk 29]
```diff
@@ -0,0 +1,67 @@
+FROM ubuntu:18.04
```
Comment (Ilya-Krylov) [inline docker/Dockerfile.cpu]: I think it can be removed, since it is not used anywhere

File: mmdet/core/bbox/coder/delta_xywh_bbox_coder.py [hunk 30]
```diff
@@ -4,6 +4,8 @@
 from ..builder import BBOX_CODERS
 from .base_bbox_coder import BaseBBoxCoder
 
+from mmdet.core.bbox.transforms import clamp
```
Comment (Ilya-Krylov) [inline mmdet/core/bbox/coder/delta_xywh_bbox_coder.py]: unused

File: mmdet/core/bbox/transforms.py [hunk 31]
```diff
@@ -1,5 +1,37 @@
 import numpy as np
 import torch
+from torch.onnx import is_in_onnx_export
+
+from ..utils.misc import to_numpy
```
Comment (Ilya-Krylov) [inline mmdet/core/bbox/transforms.py]: unused

File: .github/workflows/build.yml (hunk 1 above)
Comment (DmitriySidnev) [reply .github/workflows/build.yml]: I tried to install extensions with `cuda` in tests. Removed now.

File: configs/albu_example/mask_rcnn_r50_fpn_albu_1x_coco.py (hunk 3 above)
Comment (DmitriySidnev) [reply configs/albu_example/mask_rcnn_r50_fpn_albu_1x_coco.py:34]: Due to the deprecation warning.

File: mmdet/core/anchor/anchor_generator.py (hunk 9 above)
Comment (DmitriySidnev) [reply mmdet/core/anchor/anchor_generator.py]: Removed.

File: mmdet/core/bbox/assigners/atss_assigner.py (hunk 10 above)
Comment (DmitriySidnev) [reply mmdet/core/bbox/assigners/atss_assigner.py]: Removed.

File: mmdet/core/post_processing/bbox_nms.py (hunk 11 above)
Comment (DmitriySidnev) [reply mmdet/core/post_processing/bbox_nms.py]: Done.

File: mmdet/core/post_processing/bbox_nms.py (hunk 12 above)
Comment (DmitriySidnev) [reply mmdet/core/post_processing/bbox_nms.py]: Removed.

Comment (druzhkov-paul) [reply mmdet/core/post_processing/bbox_nms.py]: Indeed?

File: mmdet/datasets/pipelines/loading.py (hunk 14 above)
Comment (DmitriySidnev) [reply mmdet/datasets/pipelines/loading.py]: Fixed.

File: mmdet/datasets/utils.py (hunk 16 above)
Comment (DmitriySidnev) [reply mmdet/datasets/utils.py:130]: Backbones from `pytorchcv` have class field `num_classes` while local backbones do not have this field. At least it leads to fail in test with `NumClassCheckHook`.

File: mmdet/models/dense_heads/fcos_head.py (hunk 17 above)
Comment (DmitriySidnev) [reply mmdet/models/dense_heads/fcos_head.py]: Fixed to the upstream implementation.

File: mmdet/models/dense_heads/fovea_head.py (hunk 18 above)
Comment (DmitriySidnev) [reply mmdet/models/dense_heads/fovea_head.py]: Fixed to the upstream implementation.

File: mmdet/models/detectors/base.py (hunk 20 above)
Comment (DmitriySidnev) [reply mmdet/models/detectors/base.py]: Cleaned up.

File: mmdet/models/roi_heads/htc_roi_head.py (hunk 21 above)
Comment (DmitriySidnev) [reply mmdet/models/roi_heads/htc_roi_head.py]: It works only in case if target feature map size is 2 times smaller. With `adaptive_avg_pool2d` the model is not exported to ONNX. So, I removed this patch and suggest to work on it later.

File: mmdet/models/roi_heads/mask_heads/fcn_mask_head.py (hunk 22 above)
Comment (DmitriySidnev) [reply mmdet/models/roi_heads/mask_heads/fcn_mask_head.py]: No. Removed.

File: requirements/optional.txt (hunk 25 above)
Comment (DmitriySidnev) [reply requirements/optional.txt]: Probably mistakenly. Reverted.

File: setup.py (hunk 26 above)
Comment (DmitriySidnev) [reply setup.py]: Suggest to remain it as is now and do it later.

File: mmdet/core/bbox/coder/delta_xywh_bbox_coder.py (hunk 30 above)
Comment (DmitriySidnev) [reply mmdet/core/bbox/coder/delta_xywh_bbox_coder.py]: Yes. Removed.

File: mmdet/models/roi_heads/mask_heads/fcn_mask_head.py (hunk 22 above)
Comment (druzhkov-paul) [reply mmdet/models/roi_heads/mask_heads/fcn_mask_head.py]: For all that it seems to be used. Please return import as it's done in upstream.

File: setup.py (hunk 26 above)
Comment (druzhkov-paul) [reply setup.py]: Why? What makes it being out of this PR's scope?

File: mmdet/models/detectors/base.py (hunk 19 above)
Comment (DmitriySidnev) [reply mmdet/models/detectors/base.py:12]: As in [ote](https://github.com/openvinotoolkit/mmdetection/blob/ote/mmdet/models/detectors/base.py) branch. What exactly confuses you?

File: .isort.cfg (hunk 2 above)
Comment (DmitriySidnev) [reply .isort.cfg]: `.isort.cfg` removed.

File: configs/fcos/fcos_center_r50_caffe_fpn_gn-head_4x4_1x_coco.py (hunk 4 above)
Comment (DmitriySidnev) [reply configs/fcos/fcos_center_r50_caffe_fpn_gn-head_4x4_1x_coco.py]: Removed. Also fixed tests.

File: docs/conf.py (hunk 5 above)
Comment (DmitriySidnev) [reply docs/conf.py]: Restored.

File: mmdet/apis/test.py (hunk 6 above)
Comment (DmitriySidnev) [reply mmdet/apis/test.py]: Yes. Reverted.

File: mmdet/core/anchor/anchor_generator.py (hunk 7 above)
Comment (DmitriySidnev) [reply mmdet/core/anchor/anchor_generator.py]: Cleaned up.

File: mmdet/core/post_processing/bbox_nms.py (hunk 12 above)
Comment (DmitriySidnev) [reply mmdet/core/post_processing/bbox_nms.py]: Now removed.

File: mmdet/core/utils/lr_updater.py (hunk 13 above)
Comment (DmitriySidnev) [reply mmdet/core/utils/lr_updater.py]: Done.

File: mmdet/datasets/samplers/distributed_sampler.py (hunk 15 above)
Comment (DmitriySidnev) [reply mmdet/datasets/samplers/distributed_sampler.py]: Fixed to the upstream implementation.

File: .github/CONTRIBUTING.md (hunk 27 above)
Comment (DmitriySidnev) [reply .github/CONTRIBUTING.md:1]: Done as you suggested.

File: README.md (hunk 28 above)
Comment (DmitriySidnev) [reply README.md]: Added.

File: docker/Dockerfile.cpu (hunk 29 above)
Comment (DmitriySidnev) [reply docker/Dockerfile.cpu]: `Dockerfile.cpu` removed.

File: mmdet/core/bbox/transforms.py (hunk 31 above)
Comment (DmitriySidnev) [reply mmdet/core/bbox/transforms.py]: Removed.

File: mmdet/models/roi_heads/mask_heads/fcn_mask_head.py (hunk 22 above)
Comment (DmitriySidnev) [reply mmdet/models/roi_heads/mask_heads/fcn_mask_head.py]: There was double import of `CARAFEPack` and one was deleted.

File: setup.py (hunk 26 above)
Comment (DmitriySidnev) [reply setup.py]: As I remember with extensions implemented in `mmcv` our python code does not work (only deformable convolutions work). I will check it again.

File: mmdet/core/bbox/coder/delta_xywh_bbox_coder.py [hunk 32]
```diff
@@ -195,6 +195,7 @@ def delta2bbox(rois,
     y1 = gy - gh * 0.5
     x2 = gx + gw * 0.5
     y2 = gy + gh * 0.5
+
```
Comment (Ilya-Krylov) [inline mmdet/core/bbox/coder/delta_xywh_bbox_coder.py]: please remove useless change :)

File: mmdet/models/backbones/__init__.py [hunk 33]
```diff
@@ -1,3 +1,4 @@
+from .mobilenetv2 import SSDMobilenetV2
```
Comment (Ilya-Krylov) [inline mmdet/models/backbones/__init__.py]: let's remove SSDMobilenetV2 from repo

File: mmdet/models/backbones/mobilenetv2.py [hunk 34]
```diff
@@ -0,0 +1,255 @@
+# Copyright (C) 2020 Intel Corporation
```
Comment (Ilya-Krylov) [inline mmdet/models/backbones/mobilenetv2.py]: this file can be removed I guess.

File: tools/config_py_to_yaml_or_json.py [hunk 35]
```diff
@@ -0,0 +1,48 @@
+import os.path as osp
```
Comment (Ilya-Krylov) [inline tools/config_py_to_yaml_or_json.py]: please remove this file

File: mmdet/utils/collect_env.py [hunk 36]
```diff
@@ -7,7 +9,33 @@
 def collect_env():
     """Collect the information of the running environments."""
     env_info = collect_base_env()
-    env_info['MMDetection'] = mmdet.__version__ + '+' + get_git_hash()[:7]
+    env_info['MMDetection'] = mmdet.__version__
```
Comment (druzhkov-paul) [inline mmdet/utils/collect_env.py]:
```suggestion
    env_info['MMDetection'] = mmdet.__version__ + '+' + get_git_hash()[:7]
```

File: mmdet/core/bbox/coder/delta_xywh_bbox_coder.py (hunk 32 above)
Comment (DmitriySidnev) [reply mmdet/core/bbox/coder/delta_xywh_bbox_coder.py]: Fixed.

File: mmdet/models/backbones/__init__.py (hunk 33 above)
Comment (DmitriySidnev) [reply mmdet/models/backbones/__init__.py]: Removed.

File: mmdet/models/backbones/mobilenetv2.py (hunk 34 above)
Comment (DmitriySidnev) [reply mmdet/models/backbones/mobilenetv2.py]: Removed.

File: tools/config_py_to_yaml_or_json.py (hunk 35 above)
Comment (DmitriySidnev) [reply tools/config_py_to_yaml_or_json.py]: Removed.

File: mmdet/models/detectors/base.py (hunk 19 above)
Comment (druzhkov-paul) [reply mmdet/models/detectors/base.py:12]: The main that I care at this point is diff to upstream -- it's better to make it as small as possible. while keeping all the added features functional. Though, as soon as we have lots of other linter issues with our code, you may leave it as is.

File: setup.py (hunk 26 above)
Comment (DmitriySidnev) [reply setup.py]: Local `RoIAlign` was removed.

**Comment by DmitriySidnev** [comment]
run tests

File: mmdet/core/utils/lr_updater.py [hunk 37]
```diff
@@ -0,0 +1,42 @@
...
+        last_epoch (int, optional): The number of the last epoch where LR
+            updating stops. This value can be greater than total_epochs. If it
+            is equal to -1 LR will update until the end of the training.
+            Default: -1.
+    """
+
+    def __init__(self, last_epoch=-1, **kwargs):
+        assert last_epoch != 0
+        self.last_epoch = int(last_epoch)
+        super(CosineAnealingUntilEpochLrUpdaterHook, self).__init__(**kwargs)
```
Comment (druzhkov-paul) [inline mmdet/core/utils/lr_updater.py]:
```suggestion
        super().__init__(**kwargs)
```

File: configs/efficientdet/retinanet_effd0_bifpn_1x_coco.py [hunk 38]
```diff
@@ -0,0 +1,165 @@
...
+        img_prefix=data_root + 'val2017',
+        test_mode=True,
+        pipeline=test_pipeline))
+evaluation = dict(interval=1, metric='bbox')
+# optimizer
+optimizer = dict(type='SGD', lr=0.08, momentum=0.9, weight_decay=4e-5)
+optimizer_config = dict(grad_clip=None)
+# learning policy
+lr_config = dict(
+    policy='CosineAnealing',
```
Comment (druzhkov-paul) [inline configs/efficientdet/retinanet_effd0_bifpn_1x_coco.py]:
```suggestion
    policy='CosineAnnealing',
```

File: mmdet/core/utils/lr_updater.py [hunk 39]
```diff
@@ -0,0 +1,42 @@
+from mmcv.runner import HOOKS
+from mmcv.runner.hooks.lr_updater import CosineAnnealingLrUpdaterHook, annealing_cos
+
+
+@HOOKS.register_module()
+class CosineAnnealingUntilEpochLrUpdaterHook(CosineAnnealingLrUpdaterHook):
+    """The same LR updater as CosineAnealing but with `last epoch` support.
```
Comment (druzhkov-paul) [inline mmdet/core/utils/lr_updater.py]:
```suggestion
    """The same LR updater as CosineAnnealing but with `last epoch` support.
```
